File: app/src/main/java/com/elseboot3909/GCRClient/UI/Main/ChangesListFragment.java
package com.elseboot3909.GCRClient.UI.Main;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.GestureDetector;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.elseboot3909.GCRClient.API.ChangesAPI;
import com.elseboot3909.GCRClient.Adapter.ChangePreviewAdapter;
import com.elseboot3909.GCRClient.Entities.ChangeInfo;
import com.elseboot3909.GCRClient.R;
import com.elseboot3909.GCRClient.UI.Change.ChangeActivity;
import com.elseboot3909.GCRClient.UI.Search.SearchActivity;
import com.elseboot3909.GCRClient.Utils.Constants;
import com.elseboot3909.GCRClient.Utils.JsonUtils;
import com.elseboot3909.GCRClient.Utils.NetManager;
import com.elseboot3909.GCRClient.databinding.FragmentChangesListBinding;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.util.ArrayList;
import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class ChangesListFragment extends Fragment {

    FragmentChangesListBinding binding;
    private final ArrayList<ChangeInfo> changesList = new ArrayList<>();
    private ArrayList<String> queryParams = new ArrayList<>();
    private ChangePreviewAdapter changePreviewAdapter;
    private GestureDetector mGestureDetector;
    private boolean isFrozen = false;

    ActivityResultLauncher<Intent> activityResultLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Constants.SEARCH_ACQUIRED) {
                    queryParams.clear();
                    String search = result.getData().getStringExtra("search_string");
                    if (!search.isEmpty()) {
                        queryParams.add(search);
                        binding.searchBar.setText(search);
                    } else {
                        queryParams.add("status:open");
                        binding.searchBar.setText("");
                    }
                    int oldSize = changesList.size();
                    changesList.clear();
                    changePreviewAdapter.notifyItemRangeRemoved(0, oldSize);
                }
            }
    );

    public ChangesListFragment() { }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        binding = FragmentChangesListBinding.inflate(inflater, container, false);

        binding.ctlMenu.setOnClickListener((view) -> {
            DrawerLayout serverNavDL = Objects.requireNonNull(getActivity()).findViewById(R.id.serverNavDL);
            serverNavDL.openDrawer(GravityCompat.START);
        });

        if (queryParams.isEmpty()) {
            queryParams.add("status:open");
        }
        binding.searchBar.setOnClickListener(view -> {
            Intent intent = new Intent(getActivity(), SearchActivity.class);
            intent.putExtra("search_string", binding.searchBar.getText().toString().trim());
            activityResultLauncher.launch(intent);
        });

        if (changesList.isEmpty()) {
            getChangesList();
        }

        changePreviewAdapter = new ChangePreviewAdapter(changesList, this);
        binding.changesView.setAdapter(changePreviewAdapter);
        binding.changesView.setLayoutManager(new LinearLayoutManager(getContext()));
        binding.changesView.setNestedScrollingEnabled(true);

        binding.changesView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (!recyclerView.canScrollVertically(1)){
                    getChangesList();
                }
            }
        });

        mGestureDetector = new GestureDetector(getContext(), new GestureDetector.SimpleOnGestureListener() {
            @Override
            public boolean onSingleTapUp(MotionEvent e) {
                return true;
            }
        });

        binding.changesView.addOnItemTouchListener(new RecyclerView.OnItemTouchListener() {
            @Override
            public boolean onInterceptTouchEvent(@NonNull RecyclerView rv, @NonNull MotionEvent e) {
                View child = rv.findChildViewUnder(e.getX(), e.getY());
                if (mGestureDetector.onTouchEvent(e) && child != null) {
                    ChangeInfo changeInfo = changesList.get(rv.getChildAdapterPosition(child));
                    Intent intent = new Intent(getActivity(), ChangeActivity.class);
                    intent.putExtra("id", changeInfo.getId());
                    intent.putExtra("subject", changeInfo.getSubject());
                    intent.putExtra("starred", changeInfo.getStarred());
                    intent.putExtra("status", changeInfo.getStatus());
                    intent.putExtra("work_in_progress", changeInfo.getWork_in_progress());
                    startActivity(intent);
                }
                return isFrozen;
            }

            @Override
            public void onTouchEvent(@NonNull RecyclerView rv, @NonNull MotionEvent e) { }

            @Override
            public void onRequestDisallowInterceptTouchEvent(boolean disallowIntercept) { }
        });

        binding.refreshChangesView.setOnRefreshListener(() -> {
            int oldSize = changesList.size();
            changesList.clear();
            changePreviewAdapter.notifyItemRangeRemoved(0, oldSize);
            binding.refreshChangesView.setRefreshing(false);
        });

        return binding.getRoot();
    }

    private void getChangesList() {
        ((MainActivity) Objects.requireNonNull(getActivity())).progressBarManager(true);
        isFrozen = true;

        int oldSize = changesList.size();

        Retrofit retrofit = NetManager.getRetrofitConfiguration(null, true);

        ChangesAPI changesAPI = retrofit.create(ChangesAPI.class);
        Call<String> retrofitRequest = changesAPI.queryChanges(queryParams, 20, oldSize);

        retrofitRequest.enqueue(new Callback<String>() {
            @Override
            public void onResponse(@NonNull Call<String> call, @NonNull Response<String> response) {
                if (response.isSuccessful() && response.body() != null) {
                    Gson gson = new Gson();
                    changesList.addAll(gson.fromJson(JsonUtils.TrimJson(response.body()),
                            new TypeToken<ArrayList<ChangeInfo>>() {
                    }.getType()));
                    Log.e(Constants.LOG_TAG, String.valueOf(changesList.size()));
                    changePreviewAdapter.notifyItemRangeChanged(oldSize, 20);
                } else {
                    Log.e(Constants.LOG_TAG, "onResponse: Not successful");
                }
                ((MainActivity) Objects.requireNonNull(getActivity())).progressBarManager(false);
                isFrozen = false;
            }

            @Override
            public void onFailure(@NonNull Call<String> call, @NonNull Throwable t) {
                ((MainActivity) Objects.requireNonNull(getActivity())).progressBarManager(false);
                isFrozen = false;
                Log.e(Constants.LOG_TAG, "onFailure: Not successful");
            }
        });
    }

}

File: app/src/main/java/com/elseboot3909/GCRClient/UI/Main/ChangesListFragment.kt
package com.elseboot3909.GCRClient.UI.Main

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.*
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.elseboot3909.GCRClient.API.ChangesAPI
import com.elseboot3909.GCRClient.Adapter.ChangePreviewAdapter
import com.elseboot3909.GCRClient.Entities.ChangeInfo
import com.elseboot3909.GCRClient.R
import com.elseboot3909.GCRClient.UI.Change.ChangeActivity
import com.elseboot3909.GCRClient.UI.Search.SearchActivity
import com.elseboot3909.GCRClient.Utils.Constants
import com.elseboot3909.GCRClient.Utils.JsonUtils
import com.elseboot3909.GCRClient.Utils.NetManager
import com.elseboot3909.GCRClient.databinding.FragmentChangesListBinding
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.Serializable

class ChangesListFragment : Fragment() {

    private lateinit var binding: FragmentChangesListBinding
    private var changesList = ArrayList<ChangeInfo>()
    private var queryParams = ArrayList<String>()
    private val mGestureDetector = GestureDetector(context, object: GestureDetector.SimpleOnGestureListener() {
        override fun onSingleTapUp(e: MotionEvent) : Boolean {
            return true
        }
    })
    private lateinit var changePreviewAdapter: ChangePreviewAdapter
    private var isFrozen = false

    private val activityResultLauncher = registerForActivityResult(
            ActivityResultContracts.StartActivityForResult()
    ) { result ->
        when (result.resultCode) {
            Constants.SEARCH_ACQUIRED -> {
                queryParams.clear()
                val search = result.data?.getStringExtra("search_string")
                if (search?.isNotEmpty() == true) {
                    queryParams.add(search)
                    binding.searchBar.setText(search)
                } else {
                    queryParams.add("status:open")
                    binding.searchBar.setText("")
                }
                val oldSize = changesList.size
                changesList.clear()
                changePreviewAdapter.notifyItemRangeRemoved(0, oldSize)
            }
            Constants.CHANGE_STATE_CHANGED -> {
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = FragmentChangesListBinding.inflate(inflater, container, false)


        activity?.let {
            changePreviewAdapter = ChangePreviewAdapter(changesList, it)
        }
        binding.changesView.adapter = changePreviewAdapter
        binding.changesView.layoutManager = LinearLayoutManager(context)
        binding.changesView.isNestedScrollingEnabled = true

        binding.ctlMenu.setOnClickListener {
            activity?.findViewById<DrawerLayout>(R.id.serverNavDL)?.openDrawer(GravityCompat.START)
        }

        if (queryParams.isEmpty()) {
            queryParams.add("status:open")
        }
        binding.searchBar.setOnClickListener {
            val intent = Intent(activity, SearchActivity::class.java)
            intent.putExtra("search_string", binding.searchBar.text.toString().trim())
            activityResultLauncher.launch(intent)
        }

        if (changesList.isEmpty()) {
            getChangesList()
        }

        binding.changesView.addOnScrollListener(object: RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (!recyclerView.canScrollVertically(1)){
                    getChangesList()
                }
            }
        })

        binding.changesView.addOnItemTouchListener(object: RecyclerView.OnItemTouchListener {
            override fun onInterceptTouchEvent(rv: RecyclerView, e: MotionEvent): Boolean {
                val child = rv.findChildViewUnder(e.x, e.y)
                if (mGestureDetector.onTouchEvent(e) && child != null) {
                    val changeInfo = changesList[rv.getChildAdapterPosition(child)]
                    val intent = Intent(activity, ChangeActivity::class.java)
                    intent.putExtra("changeInfo", changeInfo as Serializable)
                    startActivity(intent)
                }
                return isFrozen
            }

            override fun onTouchEvent(rv: RecyclerView, e: MotionEvent) { }

            override fun onRequestDisallowInterceptTouchEvent(disallowIntercept: Boolean) { }
        })

        binding.refreshChangesView.setOnRefreshListener {
            val oldSize = changesList.size
            changesList.clear()
            changePreviewAdapter.notifyItemRangeRemoved(0, oldSize)
            binding.refreshChangesView.isRefreshing = false
        }

        return binding.root
    }

    private fun getChangesList() {
        (activity as MainActivity).progressBarManager(true)
        isFrozen = true

        val oldSize = changesList.size

        val retrofit = NetManager.getRetrofitConfiguration(null, true)
        val changesAPI = retrofit.create(ChangesAPI::class.java)

        changesAPI.queryChanges(queryParams, 20, oldSize).enqueue(object : Callback<String> {

            override fun onResponse(call: Call<String>, response: Response<String>) {
                if (response.isSuccessful && response.body() != null) {
                    changesList.addAll(Gson().fromJson(JsonUtils.trimJson(response.body()), object : TypeToken<ArrayList<ChangeInfo>>() {}.type))
                    changePreviewAdapter.notifyItemRangeChanged(oldSize, 20)
                } else {
                    Log.e(Constants.LOG_TAG, "onResponse: Not successful")
                }
                (activity as MainActivity).progressBarManager(false)
                isFrozen = false
            }

            override fun onFailure(call: Call<String>, t: Throwable) {
                (activity as MainActivity).progressBarManager(false)
                isFrozen = false
                Log.e(Constants.LOG_TAG, "onFailure: Not successful")
            }
        })
    }

}